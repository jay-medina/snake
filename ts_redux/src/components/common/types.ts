export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Snake = GridItem[];

export type GameState = 'start' | 'run' | 'gameover';
