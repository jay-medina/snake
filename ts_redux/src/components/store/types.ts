import {
  MapDispatchToProps as MDTP,
  MapStateToProps as MSTP,
} from 'react-redux';
import { AppState } from './util';

export type MapDispatchToProps<DispatchProps, OwnProps = {}> = MDTP<
  DispatchProps,
  OwnProps
>;

export type MapStateToProps<StateProps, OwnProps = {}> = MSTP<
  StateProps,
  OwnProps,
  AppState
>;
