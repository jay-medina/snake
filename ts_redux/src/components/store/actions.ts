// import { GameThunkAction } from './types';
// import { AppState } from './util';

interface GameAction {
  type: 'START_GAME';
}

export type AppAction = GameAction;

export const startGame = (): GameAction => ({
  type: 'START_GAME',
});

// export function randomizeApple(appState: AppState) {

// }

// export const startGameThunk = (): GameThunkAction => (dispatch) => {
//   // randomize the apple position
//   // start the game
// //  return dispatch(startGame);
// };
