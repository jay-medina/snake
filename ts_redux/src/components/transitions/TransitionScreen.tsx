import React, { SFC } from 'react';
import { GameState } from '../common/types';
import { PlayButton } from './PlayButton';

interface TransitionScreenProps {
  gameState: GameState;
  onPlayClick(): void;
}

interface GameOverProps {
  onPlayClick: () => void;
}

interface StartProps {
  onPlayClick(): void;
}

const Start = ({ onPlayClick }: StartProps) => (
  <div className="snake__start-screen">
    <div className="snake__start-screen-title">Snake</div>
    <PlayButton onClick={onPlayClick} />
  </div>
);

const GameOver = ({ onPlayClick }: GameOverProps) => {
  return (
    <div className="snake__game-over-screen">
      <div className="snake__game-over-title">Game Over</div>
      <PlayButton onClick={onPlayClick} />
    </div>
  );
};

export const TransitionScreen: SFC<TransitionScreenProps> = ({
  gameState,
  onPlayClick,
}) => {
  if (gameState === 'start') {
    return <Start onPlayClick={onPlayClick} />;
  }
  if (gameState === 'gameover') {
    return <GameOver onPlayClick={onPlayClick} />;
  }

  return null;
};

TransitionScreen.displayName = 'TransitionScreen';
