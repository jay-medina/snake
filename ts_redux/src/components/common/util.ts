import { GridItem, Snake } from 'store/util';

const isAtPosition = (gridItem1: GridItem) => (gridItem2: GridItem) => {
  return gridItem1.row === gridItem2.row && gridItem1.col === gridItem2.col;
};

export const isSnakeAtPosition = (snake: Snake) => (gridItem: GridItem) =>
  !!snake.find(isAtPosition(gridItem));

export const isTheApple = isAtPosition;
