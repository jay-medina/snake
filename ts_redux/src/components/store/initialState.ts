import { AppState } from "./util";


const loadInitialState = (): Partial<AppState> => {
  return {
    apple: 
  };
}
