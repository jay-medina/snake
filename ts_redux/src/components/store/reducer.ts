import { AppAction } from './actions';
import { AppState, GameState, initialState } from './util';

type Game = (start: AppState | undefined, action: AppAction) => AppState;

export const game: Game = (state = initialState, action) => {
  if (action.type === 'START_GAME') {
    return {
      gameState: GameState.Run,
    };
  }

  return state;
};
