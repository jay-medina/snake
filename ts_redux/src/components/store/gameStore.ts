import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { game } from './reducer';

export function createGameStore() {
  return createStore(game, applyMiddleware(thunk));
}
