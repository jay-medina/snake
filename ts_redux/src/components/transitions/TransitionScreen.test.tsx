jest.mock('../../common/util');

import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import { AppState } from '../../common/types';
import TransitionScreen from './TransitionScreen';
import { Provider } from 'react-redux';
import { startGame, startGameThunk } from '../../store/actions';
import {
  createMockState,
  createTestStore,
  StoreCreator,
  MockStore,
} from '../../common/testhelpers';

type noop = () => void;

describe('<TransitionScreen />', () => {
  let wrapper: ReactWrapper;
  let mockState: AppState;
  let storeCreator: StoreCreator;
  let store: MockStore;

  beforeEach(() => {
    mockState = createMockState();
    storeCreator = createTestStore();
  });

  describe('when the game is running', () => {
    beforeEach(() => {
      store = storeCreator({
        ...mockState,
        gameState: 'Run',
      });

      wrapper = mount(
        <Provider store={store}>
          <TransitionScreen />
        </Provider>,
      );
    });

    it('doesnt render a screen', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('passes the gameState to the Transition Screen', () => {
      expect(wrapper.find('TransitionScreen').prop('gameState')).toBe('Run');
    });
  });

  describe('when the game is in start state', () => {
    beforeEach(() => {
      store = storeCreator({
        ...mockState,
        gameState: 'Start',
      });

      wrapper = mount(
        <Provider store={store}>
          <TransitionScreen />
        </Provider>,
      );
    });

    it('renders the start screen', () => {
      expect(wrapper.find('Start')).toMatchSnapshot();
    });

    describe('when user clicks play', () => {
      beforeEach(() => {
        (startGameThunk as jest.Mock) = jest.fn(startGame);

        const onPlayClick = wrapper.find('Start').prop<noop>('onPlayClick');

        onPlayClick();
      });

      it('dispatches the start game thunk action', () => {
        expect(startGameThunk).toHaveBeenCalled();
        expect(store.getActions()).toContainEqual(startGame());
      });
    });
  });

  describe('when the game is in end state', () => {
    beforeEach(() => {
      store = storeCreator({
        ...mockState,
        gameState: 'GameOver',
      });

      wrapper = mount(
        <Provider store={store}>
          <TransitionScreen />
        </Provider>,
      );
    });

    it('renders the end screen', () => {
      expect(wrapper.find('GameOver')).toMatchSnapshot();
    });

    describe('when user clicks play', () => {
      beforeEach(() => {
        (startGameThunk as jest.Mock) = jest.fn(startGame);

        const onPlayClick = wrapper.find('GameOver').prop<noop>('onPlayClick');

        onPlayClick();
      });

      it('dispatches the start game thunk action', () => {
        expect(startGameThunk).toHaveBeenCalled();
        expect(store.getActions()).toContainEqual(startGame());
      });
    });
  });
});
