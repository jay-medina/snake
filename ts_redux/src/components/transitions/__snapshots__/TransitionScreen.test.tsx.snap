// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransitionScreen /> when the game is in end state renders the end screen 1`] = `
<GameOver
  onPlayClick={[Function]}
>
  <div
    className="snake__game-over-screen"
  >
    <div
      className="snake__game-over-title"
    >
      Game Over
    </div>
    <PlayButton
      onClick={[Function]}
    >
      <div
        className="snake__start-screen-play-container"
      >
        <button
          className="snake__start-screen-play"
          onClick={[Function]}
        >
          Play
        </button>
      </div>
    </PlayButton>
  </div>
</GameOver>
`;

exports[`<TransitionScreen /> when the game is in start state renders the start screen 1`] = `
<Start
  onPlayClick={[Function]}
>
  <div
    className="snake__start-screen"
  >
    <div
      className="snake__start-screen-title"
    >
      Snake
    </div>
    <PlayButton
      onClick={[Function]}
    >
      <div
        className="snake__start-screen-play-container"
      >
        <button
          className="snake__start-screen-play"
          onClick={[Function]}
        >
          Play
        </button>
      </div>
    </PlayButton>
  </div>
</Start>
`;

exports[`<TransitionScreen /> when the game is running doesnt render a screen 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TransitionScreen)>
    <TransitionScreen
      gameState={1}
      onPlayClick={[Function]}
    />
  </Connect(TransitionScreen)>
</Provider>
`;
